pg_search/src/postgres/rel.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2023-2025 ParadeDB, Inc.
+//
+// This file is part of ParadeDB - Postgres for Search and Analytics
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//! Provides a reference-counted wrapper around an open Postgres [`pg_sys::Relation`].
+use pgrx::{name_data_to_str, pg_sys, PgList, PgTupleDesc};
+use std::fmt::{Debug, Formatter};
+use std::ops::Deref;
+use std::ptr::NonNull;
+use std::rc::Rc;
+
+type NeedClose = bool;
+
+/// Represents an opened Postgres relation to be used by pg_search.
+///
+/// [`PgSearchRelation`] is reference counted and will close the underlying
+/// [`pg_sys::Relation`] when the last reference is dropped, accounting for
+/// the state of the current transaction.
+///
+/// Instances of [`PgSearchRelation`] can be closed as necessary.
+#[allow(clippy::type_complexity)]
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct PgSearchRelation(
+    Option<
+        Rc<(
+            NonNull<pg_sys::RelationData>,
+            NeedClose,
+            Option<pg_sys::LOCKMODE>,
+        )>,
+    >,
+);
+
+impl Debug for PgSearchRelation {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PgSearchRelation")
+            .field("relation", &self.oid())
+            .field("lockmode", &self.lockmode())
+            .finish()
+    }
+}
+
+impl Drop for PgSearchRelation {
+    fn drop(&mut self) {
+        if let Some(rc) = self.0.take() {
+            if let Some((relation, need_close, lockmode)) = Rc::into_inner(rc) {
```

Comment:
Nit: for guard cases like this, I like using `if-let` to early return:
```suggestion
        let Some(rc) = self.0.take() else return;
        let Some((relation, need_close, lockmode)) = Rc::into_inner(rc) else return;
```